File: Easy/1356. Sort Integers by The Number of 1 Bits.java
import java.util.Arrays;

class Solution {
    private int countBits(int num) {
        int twos = 1;
        int cnt = 0;
        while (num >= twos) {
            twos *= 2;
        }
        for (int div = twos / 2; div > 0 && num > 0; div /= 2) {
            if (num - div >= 0) {
                num -= div;
                cnt++;
            }
        }
        return cnt;
    }

    public int[] sortByBits(int[] arr) {
        // List<Integer>[] al = new ArrayList[14];
        // for (int num : arr) {
        // int cnt = countBits(num);
        // if (al[cnt] == null) {
        // al[cnt] = new ArrayList<>();
        // }
        // al[cnt].add(num);
        // }
        // for (int i = 0, idx = 0; i < al.length && idx < arr.length; i++) {
        // if (al[i] == null) {
        // continue;
        // }
        // al[i].sort((x, y) -> Integer.compare(x, y));
        // for (int num : al[i]) {
        // arr[idx++] = num;
        // }
        // }
        for (int i = 0; i < arr.length; i++) {
            arr[i] += 11111 * Integer.bitCount(arr[i]);
        }
        Arrays.sort(arr);
        for (int i = 0; i < arr.length; i++) {
            arr[i] %= 11111;
        }
        return arr;
    }
}

File: Easy/1356. Sort Integers by The Number of 1 Bits.py
from typing import List


class Solution:
    def sortByBits(self, arr: List[int]) -> List[int]:
        return sorted(arr, key = lambda num: (num.bit_count(), num))

class Solution1:
    def sortByBits(self, arr: List[int]) -> List[int]:
        # print(bin(10000))
        bits = [[] for _ in range(14)]
        for num in arr:
            bit = bin(num).count('1')
            bits[bit].append(num)
        res = []
        for i in range(14):
            res.extend(sorted(bits[i]))
        return res

